### ISSUE TITLE
[CUDA] Objective functions and metrics for cuda_exp

### ISSUE THREAD

**Issue by shiyu1994** [microsoft/LightGBM#5163]
## Summary

The `cuda_exp` in current master branch includes only a new tree learner version for CUDA. Objectives (boosting) and metrics (evaluation) should be implemented on CUDA.

**Comment by jameslamb** [comment]
Thanks for writing this up @shiyu1994 !

Can you add some more details here, to help others from the community who might want to help contribute to this?

For example...would it be possible to implement metrics one at a time (e.g. a PR like "add CUDA implementation of MAPE") and have LightGBM choose whether to evaluate it on the GPU or CPU? Or would this require a single PR implementing all metrics?

**Comment by shiyu1994** [comment]
> For example...would it be possible to implement metrics one at a time

Sure. That's what I have planned. Each PR should contain only one or several related objectives/metrics, to keep the PRs as small as possible.

**Comment by jameslamb** [comment]
Hey @shiyu1994 , are there and objective functions or metrics which still need to be added? If so, could you list them here?

If not, could we close this?

**Comment by shiyu1994** [comment]
Yes, there are still a few, which are marked in these two files:
https://github.com/microsoft/LightGBM/blob/16ca335b4a0d7a83be2b0e05133d0919b96bf35e/src/objective/objective_function.cpp#L46-L64
https://github.com/microsoft/LightGBM/blob/16ca335b4a0d7a83be2b0e05133d0919b96bf35e/src/metric/metric.cpp#L38-L70

**Comment by jameslamb** [comment]
Ahhh right, ok I was searching in the wrong place. Thank you.
